src/settings/environment.rs

```diff
@@ -0,0 +1,109 @@
+use std::collections::HashMap;
+use std::env;
+
+use config::{ConfigError, Source, Value};
+
+pub trait QueryEnvironment {
+    fn get_var(&self, var: &'static str) -> Result<String, std::env::VarError>;
+
+    fn empty(&self) -> Result<bool, failure::Error>;
+}
+
+#[derive(Clone, Debug)]
+pub struct Environment {
+    whitelist: Vec<&'static str>,
+}
+
+impl Environment {
+    pub fn with_whitelist(whitelist: Vec<&'static str>) -> Self {
+        Environment { whitelist }
+    }
+}
+
+impl QueryEnvironment for Environment {
+    fn get_var(&self, var: &'static str) -> Result<String, std::env::VarError> {
+        env::var(var)
+    }
+
+    fn empty(&self) -> Result<bool, failure::Error> {
+        let env = self.collect()?;
+
+        Ok(env.is_empty())
+    }
+}
+
+// Source trait implementation for use with Config::merge
+// until config crate removal is complete. This is effectively a
+// copy of the config crate's impl of Source for its Environment
+// struct, but rather than pulling the whole environment in #collect,
+// we pull in only whitelisted values, and rather than taking a custom
+// prefix, we assume a prefix of `CF_`.
+impl Source for Environment {
+    fn clone_into_box(&self) -> Box<dyn Source + Send + Sync> {
+        Box::new((*self).clone())
+    }
+
+    fn collect(&self) -> Result<HashMap<String, Value>, ConfigError> {
+        let mut m = HashMap::new();
+        let uri: String = "env".into();
+
+        for key in &self.whitelist {
+            if let Some(value) = env::var(key).ok() {
+                // remove the `CF` prefix before adding to collection
+                let prefix_pattern = "CF_";
+                let key = &key[prefix_pattern.len()..];
```

Comment:
nit: i think i prefer `let key = key.replace("CF_", "");` here given that `key` is an `&str` we can take advantage of `std::str::replace`

Reply:
either that or it may be worth adding an `if key.starts_with("CF_")` before doing the trim.

Reply:
this is ripped directly from `config::Environment`; replace will replace _anywhere_, not just at the front. unlikely to cause problems, but i believe in murphy's law.

Reply:
yeah the `.replace` probably not the best, but the `starts_with`

Reply:
we might even do some other voodoo here with not explicitly setting the consts to start with `CF_` but programmatically prepending before we do gets so we _know_ it starts with `CF_`